pkg/cmd/cli/cmd/logs.go

```diff
@@ -43,204 +38,110 @@ If the pod has only one container, the container name is optional.`
+		o.Options = &buildapi.BuildLogOptions{
+			Follow: podLogOptions.Follow,
 		}
-		return o.runLogsForBuild(build)
-
-	case "dc", "deploymentconfig", "deploymentconfigs":
-		dc, err := o.OriginClient.DeploymentConfigs(o.Namespace).Get(resourceName)
-		if err != nil {
-			return err
+	case *deployapi.DeploymentConfig:
+		o.Options = &deployapi.DeploymentLogOptions{
+			Follow: podLogOptions.Follow,
 		}
-		return o.runLogsForDeployment(dc)
-
 	default:
-		return fmt.Errorf("cannot display logs for resource type %v", resourceType)
+		return fmt.Errorf("cannot get logs")
 	}
-}
 
-// TODO I would recommend finding the Pod in this method and delegating the log call to the upstream LogsOptions
-// to take advantage of all of their extra options.
-func (o *OpenShiftLogsOptions) runLogsForBuild(build *buildapi.Build) error {
-	opts := buildapi.BuildLogOptions{
-		Follow: o.KubeLogOptions.Follow,
-		NoWait: false,
-	}
+	return nil
+}
 
-	readCloser, err := o.OriginClient.BuildLogs(build.Namespace).Get(build.Name, opts).Stream()
-	if err != nil {
+func (o OpenShiftLogsOptions) Validate() error {
+	if err := o.KubeLogOptions.Validate(); err != nil {
 		return err
 	}
-	defer readCloser.Close()
-
-	_, err = io.Copy(o.KubeLogOptions.Out, readCloser)
-	return err
-}
-
-func (o *OpenShiftLogsOptions) runLogsForDeployment(dc *deployapi.DeploymentConfig) error {
-	opts := deployapi.DeploymentLogOptions{
-		Follow: o.KubeLogOptions.Follow,
-		NoWait: false,
+	if o.Delegate {
+		return nil
 	}
+	// TODO: Validate our own options
+	return nil
+}
 
-	readCloser, err := o.OriginClient.DeploymentLogs(dc.Namespace).Get(dc.Name, opts).Stream()
-	if err != nil {
-		return err
+func (o OpenShiftLogsOptions) RunLog() error {
+	if !o.Delegate {
+		// Use our own options object
+		o.KubeLogOptions.Options = o.Options
 	}
-	defer readCloser.Close()
-
-	written, err := io.Copy(o.KubeLogOptions.Out, readCloser)
+	b, err := o.KubeLogOptions.RunLog()
 	if err != nil {
 		return err
 	}
-
-	// If nothing is written, it means there are no logs returned. Normally
-	// in two cases we will get no logs: either if we don't want to wait
-	// for the deployer pod to be created (NoWait = true) or if the deployer
-	// pod has been deleted (successful deployment).
-	//
-	// Note that in case of manual pod deletion or deployment pruning, this may
-	// be inaccurate.
-	if written == 0 && !opts.NoWait {
-		if opts.Version == nil {
-			fmt.Fprintln(o.KubeLogOptions.Out, "Latest deployment successfully made active, no logs to show.")
-			return nil
+	switch t := o.Options.(type) {
+	case *deployapi.DeploymentLogOptions:
+		// If nothing is written, it means there are no logs returned. Normally
+		// in two cases we will get no logs: either if we don't want to wait
+		// for the deployer pod to be created (NoWait = true) or if the deployer
+		// pod has been deleted (successful deployment).
+		//
+		// Note that in case of manual pod deletion or deployment pruning, this may
+		// be inaccurate.
```

Comment:
@ironcladlou updated to get the deployment and check if it's status is complete in case of no logs